packages/runtime/src/fetchers/graphos.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  DEFAULT_UPLINKS,
+  fetchSupergraphSdlFromManagedFederation,
+} from '@graphql-tools/federation';
+import type {
+  GatewayConfigContext,
+  GatewayGraphOSManagedFederationOptions,
+} from '../index';
+import { delayInMs } from '../utils';
+
+export interface CreateGraphOSFetcherOpts {
+  graphosOpts: GatewayGraphOSManagedFederationOptions;
+  configContext: GatewayConfigContext;
+}
+
+const defaultLoadedPlacePrefix = 'GraphOS Managed Federation';
+
+export function createGraphOSFetcher({
+  graphosOpts,
+  configContext,
+}: CreateGraphOSFetcherOpts) {
+  let lastSeenId: string;
+  let lastSupergraphSdl: string;
+  let nextFetchTime: number;
+  const uplinksParam =
+    graphosOpts.upLink || process.env['APOLLO_SCHEMA_CONFIG_DELIVERY_ENDPOINT'];
+  const uplinks =
+    uplinksParam?.split(',').map((uplink) => uplink.trim()) || DEFAULT_UPLINKS;
+  const graphosLogger = configContext.logger.child('GraphOS');
+  graphosLogger.info('Using Managed Federation with uplinks: ', ...uplinks);
+  const maxRetries = graphosOpts.maxRetries || Math.max(3, uplinks.length);
+  let supergraphLoadedPlace = defaultLoadedPlacePrefix;
+  if (graphosOpts.graphRef) {
+    supergraphLoadedPlace += ` <br>${graphosOpts.graphRef}`;
+  }
+  return {
+    supergraphLoadedPlace,
+    unifiedGraphFetcher(transportContext: TransportContext) {
+      const uplinksToUse: string[] = [];
+      let retries = maxRetries;
+      function fetchSupergraphWithDelay(): Promise<string> {
+        if (nextFetchTime) {
+          const currentTime = Date.now();
+          if (nextFetchTime >= currentTime) {
+            const delay = nextFetchTime - currentTime;
+            graphosLogger.info(`Fetching supergraph with delay: ${delay}ms`);
+            nextFetchTime = 0;
+            return delayInMs(delay).then(fetchSupergraph);
+          }
+        }
+        return fetchSupergraph();
+      }
+      function fetchSupergraph(): Promise<string> {
+        if (uplinksToUse.length === 0) {
+          uplinksToUse.push(...uplinks);
+        }
+        retries--;
+        const uplinkToUse = uplinksToUse.pop();
+        const attemptLogger = graphosLogger.child(
+          `Attempt ${maxRetries - retries} - UpLink: ${uplinkToUse}`,
+        );
+        attemptLogger.debug(`Fetching supergraph`);
+        return fetchSupergraphSdlFromManagedFederation({
+          graphRef: graphosOpts.graphRef,
+          apiKey: graphosOpts.apiKey,
+          upLink: uplinkToUse,
+          lastSeenId,
+          // @ts-expect-error TODO: Fetch types mismatch
+          fetch: transportContext.fetch || configContext.fetch,
+          loggerByMessageLevel: {
+            ERROR(message) {
+              attemptLogger.error(message);
+            },
+            INFO(message) {
+              attemptLogger.info(message);
+            },
+            WARN(message) {
+              attemptLogger.warn(message);
+            },
+          },
+        }).then(
```

Comment:
Since we always return a promise, perhaps we can in this case use an `async` function ? It should have the same perfomance impact than a `then` call ?

Reply:
You're right! Changed it.